File: vol1/ch2/LeetCode82-2.py
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head==None or head.next==None:
            return head
        h=ListNode(-1)
        h.next=head
        pre=h; p=h.next
        while p!=None:
            q=p.next     		# q指向结点p的后继结点
            if q!=None and q.val==p.val:		#结点p是重复结点
                while q!=None and q.val==p.val:
                    q=q.next
                while p!=q:        			#删除所有重复的结点
                    pre.next=p.next;p=pre.next
            else:                 	 				#结点p为非重复的结点
                pre=pre.next      				#pre,p同步后移
                p=pre.next
        return h.next

File: vol1/ch2/LeetCode82-2.cpp
class Solution {
public:
	ListNode* deleteDuplicates(ListNode* head) {
		if(head==NULL || head->next==NULL)
			return head;
		ListNode*h=new ListNode(-1);
		h->next=head;
		ListNode*pre=h,*p=h->next;
		while(p!=NULL) {
			ListNode*q=p->next;     		//q指向结点p的后继结点
			if(q!=NULL && q->val==p->val) {		//结点p是重复结点
				while(q!=NULL && q->val==p->val)	//找到后面第一个非重复结点q
					q=q->next;
				while(p!=q) {
					pre->next=p->next;
					delete p;
				p=pre->next;;
				}
			}
			else {                  					//结点p为非重复的结点
				pre=pre->next;      			//pre,p同步后移
				p=pre->next;
			}
		}
		return h->next;
	}
};


